sw/host/opentitanlib/src/util/bigint.rs

```diff
@@ -0,0 +1,375 @@
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> FixedSizeBigInt<BIT_LEN, EXACT_LEN> {
+    const BYTE_LEN: usize = BIT_LEN.saturating_add(u8::BITS as usize - 1) / u8::BITS as usize;
+
+    /// Checks the bit length of the `FixedSizeBigInt`.
+    ///
+    /// Bit length of a `FixedSizeBigInt` can be at most `BIT_LEN` if `EXACT_LEN` is `false`, must
+    /// be exactly `BIT_LEN` otherwise.
+    fn check_len(self) -> Result<Self, ParseBigIntError> {
+        match (self.bit_len().cmp(&BIT_LEN), EXACT_LEN) {
+            (Ordering::Greater, _) => Err(ParseBigIntError::PosOverflow),
+            (Ordering::Equal, _) => Ok(self),
+            (Ordering::Less, true) => Err(ParseBigIntError::PosUnderflow),
+            (Ordering::Less, false) => Ok(self),
+        }
+    }
+
+    /// Creates a `FixedSizeBigInt` from little-endian bytes.
+    pub(crate) fn from_le_bytes(bytes: impl AsRef<[u8]>) -> Result<Self, ParseBigIntError> {
+        let res = Self(BigUint::from_bytes_le(bytes.as_ref()));
+        res.check_len()
+    }
+
+    /// Creates a `FixedSizeBigInt` from big-endian bytes.
+    pub(crate) fn from_be_bytes(bytes: impl AsRef<[u8]>) -> Result<Self, ParseBigIntError> {
+        let res = Self(BigUint::from_bytes_be(bytes.as_ref()));
+        res.check_len()
+    }
+
+    /// Returns the bit length.
+    ///
+    /// Bit length of `FixedSizeBigInt` is the minimum number of bits required to represent its
+    /// value. The underlying storage may be larger.
+    pub(crate) fn bit_len(&self) -> usize {
+        self.0.bits()
+    }
+
+    /// Returns the byte representation in little-endian order.
+    pub(crate) fn to_le_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes_le()
+    }
+
+    /// Returns the byte representation in big-endian order.
+    pub(crate) fn to_be_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes_be()
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> From<FixedSizeBigInt<BIT_LEN, EXACT_LEN>>
+    for BigUint
+{
+    fn from(bigint: FixedSizeBigInt<BIT_LEN, EXACT_LEN>) -> BigUint {
+        bigint.0
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> ParseInt for FixedSizeBigInt<BIT_LEN, EXACT_LEN> {
+    type FromStrRadixErr = ParseBigIntError;
+
+    fn from_str_radix(src: &str, radix: u32) -> Result<Self, Self::FromStrRadixErr> {
+        let res =
+            Self(BigUint::from_str_radix(src, radix).map_err(ParseBigIntError::ParseBigIntError)?);
+        res.check_len()
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> fmt::Display
+    for FixedSizeBigInt<BIT_LEN, EXACT_LEN>
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let s = format!("{:#0width$x}", self.0, width = Self::BYTE_LEN * 2 + 2);
+        f.pad_integral(true, "", &s)
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> fmt::LowerHex
+    for FixedSizeBigInt<BIT_LEN, EXACT_LEN>
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
```

Comment:
Thanks for pointing out the difference between `format_args!` and calling the trait directly. I updated fmt::Display and removed the rest of the styles since we don't really need them at the moment.